### ISSUE TITLE
Botch of focus return when dismissing `Modal` by clicking outside

### ISSUE THREAD

**Issue by stokesman** [WordPress/gutenberg#51722]
### Description

When a modal is dismissed, focus is expected to return to the element that opened the modal. This fails when dismissing a modal by clicking outside it.

### Step-by-step reproduction instructions

1. Start in the Post Editor
2. Open the Options menu
3. Open the Preferences modal
4. Click outside the Preferences modal to dismiss it
5. Observe the Options menu is still open and clicking or focus other elements does not cause it to close.

### Screenshots, screen recording, code snippet

Here are a couple instances of the issue but there are more uses of `Modal` that are probably affected.

## Preferences modal
https://github.com/WordPress/gutenberg/assets/9000376/0115752a-8eb0-458a-a0ee-2906c3fc0430

## Post visibility confimation dialog
This is a `ConfirmDialog` which is built on top of `Modal`. I would expect the visibility dropdown menu would stay open and the focus return to the 'private' radio.

https://github.com/WordPress/gutenberg/assets/9000376/bb131537-e52b-40eb-a822-281c2af3bb4c

### Environment info

- Gutenberg trunk running on wp-now
- Chrome
- macOS

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes
